livekit-ffi/src/server/participant.rs

```diff
@@ -0,0 +1,189 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use livekit::prelude::*;
+use std::time::Duration;
+use tokio::sync::oneshot;
+
+use crate::{
+    proto,
+    server::room::RoomInner,
+    server::{FfiHandle, FfiServer},
+    FfiError, FfiHandleId, FfiResult,
+};
+
+#[derive(Clone)]
+pub struct FfiParticipant {
+    pub handle: FfiHandleId,
+    pub participant: Participant,
+    pub room: Arc<RoomInner>,
+}
+
+impl FfiHandle for FfiParticipant {}
+
+impl FfiParticipant {
+    pub fn perform_rpc(
+        &self,
+        server: &'static FfiServer,
+        request: proto::PerformRpcRequest,
+    ) -> FfiResult<proto::PerformRpcResponse> {
+        let async_id = server.next_id();
+
+        let local = match &self.participant {
+            Participant::Local(local) => local.clone(),
+            Participant::Remote(_) => {
+                return Err(FfiError::InvalidRequest("Expected local participant".into()))
+            }
+        };
+
+        let handle = server.async_runtime.spawn(async move {
+            let result = local
+                .perform_rpc(
+                    request.destination_identity.to_string(),
+                    request.method,
+                    request.payload,
+                    request.response_timeout_ms,
+                )
+                .await;
+
+            let callback = proto::PerformRpcCallback {
+                async_id,
+                payload: result.as_ref().ok().cloned(),
+                error: result.as_ref().err().map(|error| proto::RpcError {
+                    code: error.code,
+                    message: error.message.clone(),
+                    data: error.data.clone(),
+                }),
+            };
+
+            let _ = server.send_event(proto::ffi_event::Message::PerformRpc(callback));
+        });
+        server.watch_panic(handle);
+        Ok(proto::PerformRpcResponse { async_id })
+    }
+
+    pub fn register_rpc_method(
+        &self,
+        server: &'static FfiServer,
+        request: proto::RegisterRpcMethodRequest,
+    ) -> FfiResult<proto::RegisterRpcMethodResponse> {
+        let async_id = server.next_id();
+        let method = request.method.clone();
+
+        let local = match &self.participant {
+            Participant::Local(local) => local.clone(),
+            Participant::Remote(_) => {
+                return Err(FfiError::InvalidRequest("Expected local participant".into()))
+            }
+        };
+
+        let local_participant_handle = self.handle.clone();
+        let room: Arc<RoomInner> = self.room.clone();
+        let handle = server.async_runtime.spawn(async move {
```

Comment:
any reason why this is async?

Reply:
ah yeah you're right - the first one does need to be async and i just followed the same pattern but i can remove it, and you're also right (below) that the callbacks have no useful information so I can remove them.